domain/secret/service/watcher.go

```diff
@@ -155,6 +152,269 @@ func (s *WatchableService) WatchSecretRevisionsExpiryChanges(_ context.Context,
+		return currentChanges, nil
+	}
+
+	// pushChanges pushes the secret ID to the changes slice.
+	// At the same time, any previously added obsolete revisions of this secret are removed from the slice.
+	pushChanges := func(secretID string) {
+		currentChanges = slices.DeleteFunc(currentChanges, func(s string) bool {
+			id, _ := splitSecretRevision(s)
+			return id == secretID
+		})
+		currentChanges = append(currentChanges, secretID)
+	}
+
+	w.logger.Criticalf(ctx, "secretWatcher(%v-%v): received changes %v", w.appOwners, w.unitOwners, receivcedSecretIDs)
+	for _, uriStr := range receivcedSecretIDs {
+		uri, err := coresecrets.ParseURI(uriStr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		owned, err := w.state.IsSecretOwnedBy(ctx, uri, w.appOwners, w.unitOwners)
+		w.logger.Criticalf(ctx, "secretWatcher(%v-%v): uri %q, owned %v", w.appOwners, w.unitOwners, uri, owned)
+		if err != nil && !errors.Is(err, secreterrors.SecretNotFound) {
+			return nil, errors.Trace(err)
+		}
+		if owned {
+			// It's still owned, so the event must be triggered by creation.
+			// Ensure we are tracking the secret URI.
+			w.knownSecretURIs.Add(uri.ID)
+
+			// We are only interested in a previously owned secret that has been deleted,
+			// so ignore this one and continue.
+			continue
+		}
+
+		if !owned && w.knownSecretURIs.Contains(uri.ID) {
+			// An onwed secret has been deleted, we need to notify the URI change.
+			pushChanges(uri.ID)
+
+			// No need to track this one anymore.
+			w.knownSecretURIs.Remove(uri.ID)
+		}
+	}
+	return currentChanges, nil
+}
+
+func (w *obsoleteWatcher) mergeRevisionChanges(
+	ctx context.Context, currentChanges []string, newChanges []string,
+) ([]string, error) {
+	if len(newChanges) == 0 {
+		return currentChanges, nil
+	}
+
+	w.logger.Criticalf(ctx, "revisionWatcher(%v-%v): received changes %v", w.appOwners, w.unitOwners, newChanges)
+	// We are receiving all the obsolete revision UUIDs changes in the model, so we need to filter
+	// only the ones that are owned.
+	revisionIDs, err := w.state.GetRevisionIDsForObsolete(
+		ctx, w.appOwners, w.unitOwners, newChanges...,
+	)
+	if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+
+	w.logger.Criticalf(context.Background(), "obsoleteWatcher(%v-%v): obsoleteRevisionIDs %v", w.appOwners, w.unitOwners, revisionIDs)
+	for _, revisionID := range revisionIDs {
+		id, _ := splitSecretRevision(revisionID)
+		if !w.knownSecretURIs.Contains(id) {
+			// The secret has already removed, so we don't need to notify the obsolete revision.
+			continue
+		}
+		currentChanges = append(currentChanges, revisionID)
+	}
+	return currentChanges, nil
+}
+
+func (w *obsoleteWatcher) loop() error {
+	defer close(w.out)
+
+	// To allow the initial event to be sent.
+	out := w.out
+	var changes []string
+	for {
+		// Give any incoming secret deletion events
+		// time to arrive so the revision obsolete and
+		// delete events can be squashed.
+		timeout := time.After(time.Second)
```

Comment:
We did this in mongo but here it's not necessary. We get any obsolete revision events and go to the db to check if the secret is still there. That's the mechanism to see if the event should be discarded.